src/json.ts

```diff
@@ -266,3 +272,165 @@ export function getJsonRpcIdValidator(options?: JsonRpcValidatorOptions) {
+    if (value === undefined) {
+      // Return zero for undefined, since these are omitted from JSON serialization
+      return [true, 0];
+    } else if (value === null) {
+      // Return already specified constant size for null (special object)
+      return [true, skipSizing ? 0 : JsonSize.Null];
+    }
+
+    // Check and calculate sizes for basic types
+    // eslint-disable-next-line default-case
+    switch (typeof value) {
+      case 'function':
+        return [false, 0];
+      case 'string':
+        return [
+          true,
+          skipSizing ? 0 : calculateStringSize(value) + JsonSize.Quote * 2,
+        ];
+      case 'boolean':
+        if (skipSizing) {
+          return [true, 0];
+        }
+        return [true, value ? JsonSize.True : JsonSize.False];
+      case 'number':
+        return [true, skipSizing ? 0 : calculateNumberSize(value)];
+    }
+
+    // Handle specific complex objects that can be serialized properly
+    try {
+      if (value instanceof Date) {
+        if (skipSizing) {
+          return [true, 0];
+        }
+        return [
+          true,
+          // Note: Invalid dates will serialize to null
+          isNaN(value.getDate())
+            ? JsonSize.Null
+            : JsonSize.Date + JsonSize.Quote * 2,
+        ];
+      } else if (value instanceof Boolean) {
+        if (skipSizing) {
+          return [true, 0];
+        }
+        // eslint-disable-next-line eqeqeq
+        return [true, value == true ? JsonSize.True : JsonSize.False];
+      } else if (value instanceof Number) {
+        if (skipSizing) {
+          return [true, 0];
+        }
+        return [true, value.toString().length];
+      } else if (value instanceof String) {
+        if (skipSizing) {
+          return [true, 0];
+        }
+        return [
+          true,
+          calculateStringSize(value.toString()) + JsonSize.Quote * 2,
+        ];
+      }
+    } catch (_) {
+      return [false, 0];
+    }
+
+    // If object is not plain and cannot be serialized properly,
+    // stop here and return false for serialization
+    if (!isPlainObject(value) && !Array.isArray(value)) {
+      return [false, 0];
+    }
+
+    // Handle circular objects
+    if (seenObjects.has(value)) {
+      return [false, 0];
+    }
+    seenObjects.add(value);
+
+    // Continue object decomposition
+    try {
+      return [
+        true,
+        Object.entries(value).reduce(
+          (sum, [key, nestedValue], idx, arr) => {
+            // Recursively process next nested object or primitive type
+            // eslint-disable-next-line prefer-const
+            let [valid, size] = getJsonSerializableInfo(
+              nestedValue,
+              skipSizing,
+            );
+            if (!valid) {
+              throw new Error(
+                'JSON validation did not pass. Validation process stopped.',
+              );
+            }
+            seenObjects.delete(value);
```

Comment:
Perhaps leave a comment to explain it as well.

Reply:
Let's say that a circular structure is the one that has itself for a child node (object). Hence, the check will be recursively performed on all child nodes, following the algorithm we use for everything else (object decomposition, etc.). Once an object is visited and its status (validity and size) resolved, it can be removed from a set. This will remove the false positives you mentioned previously with multiple adjacent objects. Tricky tests for all of this are added as well.

I've also added some explanation comments in the next commit.

Also, I refactored code according to your suggestion below 👇 for the `typeof` and `instanceof` with `if-else` syntax, all at once.